### DISCUSSION TITLE
Implementing STAC format over CKAN

### DISCUSSION THREAD

**Discussion by ccancellieri** [ckan/ckan#6364] [Ideas]
### STAC

I'm currently studying the [STAC](https://github.com/radiantearth/stac-spec) format and it's [extensions](https://stac-extensions.github.io) trying to find the best way to support them into CKAN

To achieve this I'd like to leverage on a mix of technologies that I've been used with success on other types (terriajs-view)
as described [here](https://github.com/ckan/ckan/discussions/6322)

The main challenge is that I really would like to give the full control of the extensions (plus defaults) to the author of the metadata so the full/deep schema of the dataset cannot be defined statically and may support runtime (edit time) extension and validation.

I'd love that the _Editors_ will be able to create a STAC schema with a form helping him, and (why not), driving him to also create the index over **solr** to better map the value matching the new plugged extensions 
[[**solr index** configuration parsing json based field](https://solr.apache.org/guide/7_4/transforming-and-indexing-custom-json.html) ]

I know let back office users define a solr indexes and different extensions for each metadata is and hazard but I would like to keep the maximum flexibility and later limit by profile etc

Indexes configurations need also a dedicated storage and better way to define and reuse existing configurations.

FYI: I'm also looking at the ckanext-spatial plugin to reuse the spatial index and more.
I love to reuse but I don't want to have constraints and dependencies at this stage, in short: to obtain a bbox I can easily leverage over ajax + wms/wfs/wcs protocols ....

Ref:
stac - https://github.com/radiantearth/stac-spec
stac extensions - https://stac-extensions.github.io
solr from json - https://solr.apache.org/guide/7_4/transforming-and-indexing-custom-json.html
Based also on discussion - https://github.com/ckan/ckan/discussions/6322

**Comment by ccancellieri** [comment]
### Mapping

### STAC
 STAC also defines some entities which will **mapped** as follow: (DRAFT 05/09/2021)

### STAC Asset
Mapped as **Resource**
https://github.com/radiantearth/stac-spec/blob/master/collection-spec/collection-spec.md#assets
https://github.com/radiantearth/stac-spec/blob/master/collection-spec/collection-spec.md#asset-object

### STAC Collection
 Mapped as (_root type_) **Dataset** (it can be seen as a **Dataset with relations***)
https://github.com/radiantearth/stac-spec/blob/master/collection-spec/collection-spec.md#stac-collection-specification-
https://github.com/radiantearth/stac-spec/blob/master/collection-spec/collection-spec.md#collection-fields

### Dataset schema:
The 1st draft of the **dataset schema** will be:
1. **default** fields (notes, name, title, etc)
2. **extra_json_extensions** a json field: dict of schema links which are required by STAC to define [extensions](https://github.com/radiantearth/stac-spec/blob/master/extensions/README.md) used
3. **extra_extensions** (a json field: dict of values defined by the extensions specified on step 2)
4. **extra_solr** (a json field: dict of values of solr indexes)
5. The **relations** to other datasets are still unclear: I don't want to use resources with url link because that is a clear parent child relation. I'd love to fix and reuse the existing ckan relationships between datasets (which at the moment looks not recommendable to use see comments over gitter)  

### STAC Item
Mapped as simple **Dataset** 
https://github.com/radiantearth/stac-spec/blob/master/item-spec/item-spec.md

### Stac Catalog
Mapped as **Group** or **Organization**
https://github.com/radiantearth/stac-spec/blob/master/catalog-spec/catalog-spec.md

jsons fields are all a json to parse to a dict (valid json) which maps type names (extensions) to each other.

### Problems I see:
* **Relations** https://github.com/radiantearth/stac-spec/blob/master/collection-spec/collection-spec.md#relation-types
We should leverage over ckan relation to keep the consistency of the links.
As far as I understood they are not supposed to be stable and usable so I'm evaluating to register listeners to understand if/when a children or a parent is moved/deleted and prevent runtime tree navigation problems.
Open issues:
- https://github.com/ckan/ckan/issues/4212
- https://github.com/ckan/ckan/issues/5273
- https://github.com/ckan/ckan/pull/5359


### References:
extensions - https://github.com/radiantearth/stac-spec/blob/master/extensions/README.md
relations - https://github.com/radiantearth/stac-spec/blob/master/collection-spec/collection-spec.md#relation-types
stac item - https://github.com/radiantearth/stac-spec/blob/master/item-spec/item-spec.md
stac catalog - https://github.com/radiantearth/stac-spec/blob/master/catalog-spec/catalog-spec.md
stac collection - https://github.com/radiantearth/stac-spec/blob/master/collection-spec/collection-spec.md
stac asset - https://github.com/radiantearth/stac-spec/blob/master/collection-spec/collection-spec.md

**Reply by loleg** [reply]
We discussed this topic today at a [workshop on standardization](https://github.com/GeoUnconference/discussions/discussions/8), and I see a few parallels between your proposal and the [ongoing discussions](https://github.com/frictionlessdata/specs/issues/86) about geospatial standards in the @frictionlessdata community that my group here is aiming to revitalise with a [Spatial Data Package](https://github.com/cividi/spatial-data-package). I don't have anything more concrete to say at the moment, except thanks for bringing this up!

**Reply by ccancellieri** [reply]
Yes I see thanks for sharing references, I tried 6 months ago to propose the adoption of the datapackage but the plugin did not worked, so we started implementing a DCAT plugin and now we are at work over a GeoDCAT implementation.
In the meanwhile all the big partners are rapidly adopting STAC as standard for their catalogs.
GEE for example has it's own STAC browser up to date public and ready to be fetched... but it is also customizing STAC with GEE extensions.

**Comment by amercader** [comment]
@ccancellieri FYI @b-weintraub has been recently working on STAC support in CKAN as well so you might be able to exchange ideas there: https://github.com/b-weintraub/ckanext-stac

**Reply by b-weintraub** [reply]
Thanks @amercader. Hi @ccancellieri, I have also been working on extending CKAN to support STAC and would love to exchange ideas.  

I've found that there's a fair amount of variation among organizations on how they utilize the STAC format.  For example, sometimes you find the highest level of the data hierarchy as a collection, and other times it's a catalog.  I like your idea of the harvester interface allowing the editor to create a STAC schema with a form or some other input to address some of these variations. Perhaps this schema could be input as a JSON in the harvester Configuration box on the harvester interface.

Here is my first attempt at implementing a STAC harvester for ckan (https://github.com/b-weintraub/ckanext-stac). It currently supports the California Forest Observatory (https://storage.googleapis.com/cfo-public/vegetation/collection.json) and OpenTOPOGRAPHY (https://portal.opentopography.org/stac/catalog.json) STACs.

One of the challenges I ran into was that the pystac library (https://pystac.readthedocs.io/en/latest/index.html) requires python 3, which is incompatible with CKAN's python 2.7. 

It would be great to build out more customizable STAC harvesting support for CKAN.

**Reply by amercader** [reply]
> requires python 3, which is incompatible with CKAN's python 2.7.

Just to note that CKAN 2.9 supports both Python 2.7 and 3.6-8

**Reply by ccancellieri** [reply]
Oh nice to meet you, welcome!

@amercader thank you to a lot.

I still have to pull this out for testing but it's amazing to see other approaches!

Well actually our main goal is to provide an editor and an API interface to create and extend metadata into ckan,

Having an harvester is a good plus, at the moment we are leveraging over geonetwork/csw (ogc) and spatial plugin (ckan).

**Comment by ccancellieri** [comment]
Hello!
 FYI I'm doing steps ahead on this here:
https://github.com/ccancellieri/ckanext-jsonschema

The implementation is still a draft but it works and I'm able to have a quite full model **imported** directly from geonetwork into ckan.

At the moment I'm focusing over ISO19139 format but the base is the same so soon I'll provide a STAC integration based on this plugin

**Comment by ccancellieri** [comment]
Due to poor support about relationships even if a draft is available for this, I'm planning to add a graphdb extension to be able to correctly track the complex hierarchy of a STAC catalog.

**Reply by andvianello** [reply]
Hi,
is there any updates on this topic? it's really interesting and relevant for metadata catalog.
Thanks,
Andrea.

**Reply by stepps00** [reply]
Hi - I wanted to point to [an interesting use case](https://github.com/OvertureMaps/data/issues/25#issuecomment-3834197081) for STAC format over CKAN from Open Net Zero. I see this issue is a few years old; is STAC format / extensions support in CKAN still a potential enhancement you're looking at implementing? Thank you!
